src/pal/src/loader/module.cpp

```diff
@@ -1295,88 +1430,82 @@ static MODSTRUCT *LOADAllocModule(void *dl_handle, LPCSTR name)
-        // even if it has one, or to leave off a prefix like "lib" even if it has one
-        // (both of these are done typically to smooth over cross-platform differences). 
-        // We try to dlopen with such variations on the original.
-        const char* const formatStrings[4] = // used with args: PAL_SHLIB_PREFIX, shortAsciiName, PAL_SHLIB_SUFFIX
+        dl_handle = dlopen(libraryNameOrPath, RTLD_LAZY);
+        if (dl_handle == nullptr && setLastError)
         {
-            "%s%s%s",     // prefix+name+suffix
-            "%.0s%s%.0s", // name
-            "%.0s%s%s",   // name+suffix
-            "%s%s%.0s",   // prefix+name
-        };
-        const int skipPrefixing = strchr(shortAsciiName, '/') != NULL; // skip prefixing if the name is actually a path
-        for (int i = 0; i < 4; i++)
-        {
-           if (skipPrefixing && (i == 0 || i == 3)) // 0th and 3rd strings include prefixes
-               continue;
-           
-            if (!dl_handle &&
-                snprintf(fullLibraryName, MAX_LONGPATH, formatStrings[i], PAL_SHLIB_PREFIX, shortAsciiName, PAL_SHLIB_SUFFIX) < MAX_LONGPATH &&
-                ((dl_handle = dlopen(fullLibraryName, RTLD_LAZY | RTLD_NOLOAD)) || 
-                 (dl_handle = dlopen(fullLibraryName, RTLD_LAZY))))
-            {
-                shortAsciiName = fullLibraryName;
-                break;
-            }
+            WARN("dlopen() failed; dlerror says '%s'\n", dlerror());
+            SetLastError(ERROR_MOD_NOT_FOUND);
         }
     }
 
-    if (!dl_handle)
+    if (dl_handle != nullptr)
     {
-        WARN("dlopen() failed; dlerror says '%s'\n", dlerror()); 
-        SetLastError(ERROR_MOD_NOT_FOUND);
-        goto done;
+        TRACE("dlopen() found module %s\n", libraryNameOrPath);
+    }
+
+    if (!isModuleListLockedByCaller)
+    {
+        UnlockModuleList();
     }
-    TRACE("dlopen() found module %s\n", shortAsciiName);
+
+    return dl_handle;
+}
+
+/*
+Function:
+    LOADRegisterLibrary [internal]
+
+    Registers a system handle to a loaded library with the module list.
+
+Parameters:
+    HMODULE dl_handle:                  System handle to the loaded library.
+    LPCSTR libraryNameOrPath:           The library that was loaded.
+    BOOL fDynamic:                      TRUE if dynamic load through LoadLibrary, FALSE if static load through RegisterLibrary.
+    bool isModuleListLockedByCaller:    True if the caller already owns the module list lock, false otherwise.
+
+Return value:
+    PAL handle to the loaded library, or nullptr upon failure (error is set via SetLastError()).
+*/
+static HMODULE LOADRegisterLibrary(HMODULE dl_handle, LPCSTR libraryNameOrPath, BOOL fDynamic, bool isModuleListLockedByCaller)
+{
+    _ASSERTE(dl_handle != nullptr);
+    _ASSERTE(libraryNameOrPath != nullptr);
+    _ASSERTE(libraryNameOrPath[0] != '\0');
+
+    if (!isModuleListLockedByCaller)
+    {
+        LockModuleList();
+    }
```

Comment:
May be better to not have this argument, and instead change the one caller to lock the list instead.